tests/account_change_password_test.go

```diff
@@ -0,0 +1,46 @@
+package tests
+
+import (
+	"testing"
+
+	"encoding/json"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+func TestChangePassword(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+	unauthedClient := deployment.Client(t, "hs1", "")
+	newPassword := "new_superuser"
+	// sytest: After changing password, can't log in with old password
+	t.Run("After changing password, can't log in with old password", func(t *testing.T) {
+		CreateDummyUser(t, unauthedClient, "cant_login_with_old_password")
+		res := unauthedClient.DoFunc(t, "POST", []string{"_matrix", "client", "r0", "account", "password"}, client.WithRawBody(json.RawMessage(`{
+				"auth": {
+					"type": "m.login.password",
+					"user": "@cant_login_with_old_password:hs1",
+					"password": "superuser"
+				},
+				"new_password": "`+newPassword+`"
+			}`)))
+
+		must.MatchResponse(t, res, match.HTTPResponse{
+			StatusCode: 200,
```

Comment:
Hi @Kegsay , I did similar things to what you said & what is done in the sytest, I created an user & then provided auth-credentials in the request body of `POST account/password`. I am still getting an 401 in place of 200, should I pass access-key in the auth body too 😕 ?

Reply:
Please always paste the error output from the CI run so it's easier for me to help. The way I do it is to:
 - Click the red X on your commit, it'll scroll to a part of the logs. I don't want the entire log output.
 - Scroll up until you start to see homeserver logs. There are start and end markers to make this clear:
```
2021/05/24 19:55:30 ============================================
2021/05/24 19:55:30 e3026ed6adb1614ebe0f46e1891380aca613caaf08975a3f5cdcba8dd52809b6 : Server logs:

[...]

2021/05/24 19:55:30 ============== e3026ed6adb1614ebe0f46e1891380aca613caaf08975a3f5cdcba8dd52809b6 : END LOGS ==============
```
- Go to the logs just before the start of the homeserver logs and grab that error.
```
=== CONT  TestChangePassword/After_changing_password,_can't_log_in_with_old_password
    account_change_password_test.go:31: MatchResponse got status 401 want 200 - http://172.17.0.1:49158/_matrix/client/r0/account/password => {"session":"ptLNVKZxYEzyqrJIgwtrhRZN","flows":[{"stages":["m.login.email.identity"]}],"params":{},"completed":["m.login.password"]}
2021/05/24 19:55:30 ============================================
2021/05/24 19:55:30 e3026ed6adb1614ebe0f46e1891380aca613caaf08975a3f5cdcba8dd52809b6 : Server logs:
```

Complement is generally pretty good with logging responses, and this immediately tells me that your request is okay (you can see `"completed":["m.login.password"]`) but for some reason it is asking for email authentication(!!!). I'll raise this with @anoadragon453

Reply:
Sorry for the confusion (it even took me a while to figure out what's going on here)! Basically Synapse will only validate your account through the usual User Interactive Authentication flows for your account (such as `m.login.password`) if you provide an `access_token`query parameter in the request to `account/password`.

I've filed an issue [here](https://github.com/matrix-org/synapse/issues/10070) in an attempt to make things clearer for the future. But for right now to use this endpoint, you'll annoyingly need to:

1. Call `/login` with the username and password to retrieve an access token.
2. Provide that access token as a query parameter in the request to `account/password`.
3. Now you should be able to pass User-Interactive Authentication by the completing the `m.login.password` stage.

Hope that makes sense!

Reply:
Hi @anoadragon453 @Kegsay , I added the acces_token param in the account/password request endpoint. However I am getting similar error logs as above in the updated request too. Is there anything I have missed 😕 ?

Reply:
As per #complement, I see no issues with the code and don't understand why it's 401ing you still. Run it against Dendrite and if it passes then add this test to synapse blacklist and move on.

Reply:
It appears Synapse isn't seeing the `Authorization` header for whatever reason. Do double-check it's being sent with `COMPLEMENT_DEBUG=1`, otherwise I'm also at a loss.

Reply:
Fixed, dendrite logs were more accurate and showed that the error was at the client side which was sending the access_token.

Reply:
However this brings an weird behaviour of complement. user_id & access_token were not getting assigned to the client in registerUser helper, moreover the code was getting stuck at this line https://github.com/matrix-org/complement/blob/cd4a83f40ff66a3c137d923855cec56e84a6c06b/internal/client/client.go#L192 & all the statements ahead of this line were not getting executed.
When I removed these lines 
```go
	must.MatchResponse(t, res, match.HTTPResponse{		
         JSON: []match.JSON{			
          match.JSONKeyTypeEqual("access_token", gjson.String),			
          match.JSONKeyTypeEqual("user_id", gjson.String),		
       },	
``` 
from my helper function, the tests started passing. Any idea why the error was occuring 😕

Reply:
The problem is that `must.MatchResponse` will consume the response body, so when you try to read from it later on it won't work. https://github.com/matrix-org/complement/blob/master/internal/must/must.go#L72